### PR TITLE
Alternative chaining syntax for Queue, issue #44

### PR DESCRIPTION
**Pull request by tshemsedinov**

### CODE REVIEW

**Comment by tshemsedinov** [comment]
ping @aqrln

**Comment by aqrln** [comment]
@tshemsedinov yeah, the notifications about all these MetaSync PRs got completely lost in my inbox, sorry. I'll review it today.

File: lib/queue.js [hunk 1]
```diff
@@ -2,6 +2,63 @@
...
+    let drain = api.common.emptiness;
+
+    const queue = (item) => {
+      if (active && count < concurrency) queue.next(item);
+      else input.push(item);
+      let timer;
+      if (wait) {
+        timer = setTimeout(() => {
+          output(new Error('Queue wait timeout'));
+        }, wait);
```
Comment (aqrln) [inline lib/queue.js]: `timer.unref()`

File: lib/queue.js [hunk 2]
```diff
@@ -2,6 +2,63 @@
...
+      }
+      if (timer) clearTimeout(timer);
```
Comment (aqrln) [inline lib/queue.js]: This will result it timer never being set.

File: lib/queue.js [hunk 3]
```diff
@@ -2,6 +2,63 @@
...
+    let output = api.common.emptiness;
+    let drain = api.common.emptiness;
+
+    const queue = (item) => {
+      if (active && count < concurrency) queue.next(item);
+      else input.push(item);
+      let timer;
+      if (wait) {
+        timer = setTimeout(() => {
+          output(new Error('Queue wait timeout'));
```
Comment (aqrln) [inline lib/queue.js]: This doesn't prevent the items waiting in the queue from being run.

File: lib/queue.js [hunk 4]
```diff
@@ -2,6 +2,63 @@
...
+    queue.pause = () => (active = false, queue);
+    queue.resume = () => (active = true, queue.next(), queue);
+
+    queue.next = (item) => {
+      let timer;
+      if (!active) return; // maybe remove
+      count++;
+      if (timeout) {
+        timer = setTimeout(() => {
+          output(new Error('Queue process timeout'));
```
Comment (aqrln) [inline lib/queue.js]: This will not prevent the result of `process` from being emitted.

File: lib/queue.js [hunk 5]
```diff
@@ -2,6 +2,63 @@
...
+        timer = setTimeout(() => {
+          output(new Error('Queue wait timeout'));
+        }, wait);
+      }
+      if (timer) clearTimeout(timer);
+    };
+
+    queue.wait = (msec) => (wait = msec, queue);
+    queue.timeout = (msec) => (timeout = msec, queue);
+    queue.throttle = (rq, msec) => (throttles.push([rq, msec]), queue);
```
Comment (aqrln) [inline lib/queue.js]: I don't see how is throttling implemented.

File: tests/queue.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
+'use strict';
```
Comment (aqrln) [inline tests/queue.js]: Tests are not being run.

File: lib/queue.js [hunk 7]
```diff
@@ -2,6 +2,63 @@
...
+    queue.resume = () => (active = true, queue.next(), queue);
+
+    queue.next = (item) => {
+      let timer;
+      if (!active) return; // maybe remove
+      count++;
+      if (timeout) {
+        timer = setTimeout(() => {
+          output(new Error('Queue process timeout'));
+        }, timeout);
```
Comment (aqrln) [inline lib/queue.js]: `timer.unref()`

File: lib/queue.js (hunk 3 above)
Comment (tshemsedinov) [inline lib/queue.js]: These lines do not have such a goal. We use `queue.next(item)` if queue have empty slots and `imput.push(item)` if it haven't

File: lib/queue.js (hunk 5 above)
Comment (tshemsedinov) [inline lib/queue.js]: It's just a stub as of now @aqrln

**Comment by tshemsedinov** [comment]
Will be implemented as a class

**Pull request closed without merging**